tmt/libraries/__init__.py

```diff
@@ -0,0 +1,186 @@
+""" Handle libraries """
+
+import re
+from typing import List, Optional, Tuple, Union
+
+import fmf
+
+import tmt
+import tmt.base
+import tmt.log
+import tmt.utils
+from tmt.utils import Path
+
+# A beakerlib identifier type, can be a string or a fmf id (with extra beakerlib keys)
+LibraryIdentifierType = Union[
+    tmt.base.RequireSimple, tmt.base.RequireFmfId, tmt.base.RequireFile]
+ImportedIdentifiersType = Optional[List[LibraryIdentifierType]]
+
+# A Library type, can be Beakerlib or File
+# undefined references are ignored due to cyclic dependencies of these files,
+# only imported in runtime where needed
+LibraryType = Union['BeakerLib', 'File']  # type: ignore[name-defined] # noqa: F821
+
+# A type for Beakerlib dependencies
+LibraryDependenciesType = Tuple[
+    List[tmt.base.Require], List[tmt.base.Require], List['LibraryType']
+    ]
+
+
+class LibraryError(Exception):
+    """ Used when library cannot be parsed from the identifier """
+
+
+class Library:
+    """
+    General library class
+
+    Used as parent for specific libraries like beakerlib and file
+    """
+
+    def __init__(
+            self,
+            *,
+            parent: Optional[tmt.utils.Common] = None,
+            logger: tmt.log.Logger) -> None:
+        """ Process the library identifier and fetch the library """
+        # Use an empty common class if parent not provided (for logging, cache)
+        self.parent = parent or tmt.utils.Common(logger=logger, workdir=True)
+
+        self.identifier: LibraryIdentifierType
+
+    @property
+    def hostname(self) -> str:
+        """ Get hostname from url or default to local """
+        if hasattr(self, 'url') and self.url:
+            matched = re.match(r'(?:git|http|https)://(.*?)/', self.url)
+            if matched:
+                return matched.group(1)
+        return 'local'
+
+    @property
+    def fmf_node_path(self) -> Path:
+        """ Path to fmf node """
+        assert hasattr(self, 'name')  # type check
+        if hasattr(self, 'path') and self.path:
+            return Path(self.path / self.name.strip('/'))
+        return Path(self.name)
+
+    def __str__(self) -> str:
+        """ Use repo/name for string representation """
+        assert hasattr(self, 'repo')
+        assert hasattr(self, 'name')  # type check
+        return f"{self.repo}{self.name}"
```

Comment:
TL;DR: it seems to me `Library` is doing work that belongs to `BeakerLib`.

These three methods look like child implementations leaking into parents. `Library` class has no `name`, `path`, or `repo` members, therefore it should not look for them - a class that adds these attributes in its own definition should override parent methods with new code, using new attributes. It's perfectly fine to raise `NotImplementedError` here if there's no default string representation, and it's left for child classes to implement the abstract method.

But, it's also possible the attributes should be defined in this class - IIUIC, both file and beakerlib subclasses define `name` and other attributes:

```python
        # File
        self.identifier = identifier
        self.format: str = identifier.type if (
            hasattr(identifier, 'type') and isinstance(identifier.type, str)) else 'file'
        self.repo: Path = Path(target_location.name)
        self.name: Path = Path("/files")
        self.pattern: List[str] = identifier.pattern if hasattr(identifier, 'pattern') else []
        self.source_location: Path = source_location
        self.target_location: Path = target_location

        # Beakerlib is more complicated, but still, there's an overlap:
            identifier = tmt.base.RequireSimple(identifier.strip())
            self.identifier = identifier
            matched = LIBRARY_REGEXP.search(identifier)
            if not matched:
                raise LibraryError
            self.parent.debug(
                f"Detected library '{identifier.to_minimal_spec()}'.", level=3)
            self.format: str = 'rpm'
            self.repo: Path = Path(matched.groups()[0])
            self.name: str = matched.groups()[1]
            self.url: Optional[str] = DEFAULT_REPOSITORY_TEMPLATE.format(repository=self.repo)
            self.path: Optional[Path] = None
            self.ref: Optional[str] = None
            self.dest: Path = Path(DEFAULT_DESTINATION)
```

This suggests many of these attributes should be defined in `Library`, even without a value:

```python
class Library:
    identifier: LibraryIdentifierType
    format: str
    repo: Path
    name: Path
```

Then these methods could rely on these attributes to be known - `hasattr()` is often a sign of something shady, especially when types of involved objects are 100% known.

```python
    @property
    def hostname(self) -> str:
        return 'local'

    @property
    def fmf_node_path(self) -> Path:
        """ Path to fmf node """
        return self.name

    def __str__(self) -> str:
        """ Use repo/name for string representation """
        return f"{self.repo}{self.name}"

    # In BeakerLib:
    url: Optional[str] = None
    path: Optional[Path] = None

    ...

    @property
    def hostname(self) -> str:
        if self.url:
            matched = re.match(r'(?:git|http|https)://(.*?)/', self.url)
            if matched:
                return matched.group(1)
        return super().hostname

    @property
    def fmf_node_path(self) -> Path:
        """ Path to fmf node """
        if self.path:
            return self.path / self.name.strip('/')
        return super().fmf_node_path
```